ccl/oidcccl: TestOIDCEnabled failed

ccl/oidcccl.TestOIDCEnabled [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11915087?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11915087?buildTab=artifacts#/) on release-23.1.11-rc @ [62ad175cbc0e9041d993ea1cc911e07d279671f5](https://github.com/cockroachdb/cockroach/commits/62ad175cbc0e9041d993ea1cc911e07d279671f5):


```
=== RUN   TestOIDCEnabled
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3ed8b395931fde05baf66bd2e4d88d97/logTestOIDCEnabled66925503
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestOIDCEnabled
    authentication_oidc_test.go:236: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/3ed8b395931fde05baf66bd2e4d88d97/logTestOIDCEnabled66925503
--- FAIL: TestOIDCEnabled (5.26s)
=== RUN   TestOIDCEnabled/login_redirect
    authentication_oidc_test.go:206: expected 302 status code but got: 400
    --- FAIL: TestOIDCEnabled/login_redirect (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/unowned
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestOIDCEnabled.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31810

```
[03:18:00] :		 [Run stress] === RUN   TestOIDCEnabled
[03:18:00] :		 [Run stress]     test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/3ed8b395931fde05baf66bd2e4d88d97/logTestOIDCEnabled66925503
[03:18:00] :		 [Run stress]     test_log_scope.go:79: use -show-logs to present logs inline
[03:18:00] :		 [Run stress] === RUN   TestOIDCEnabled/login_redirect
[03:18:00] :		 [Run stress]     authentication_oidc_test.go:206: expected 302 status code but got: 400
[03:18:00] :		 [Run stress] === CONT  TestOIDCEnabled
[03:18:00] :		 [Run stress]     authentication_oidc_test.go:236: -- test log scope end --
[03:18:00] :		 [Run stress] test logs left over in: /artifacts/tmp/_tmp/3ed8b395931fde05baf66bd2e4d88d97/logTestOIDCEnabled66925503
[03:18:00] :		 [Run stress] --- FAIL: TestOIDCEnabled (5.26s)
```

--- a/pkg/ccl/oidcccl/BUILD.bazel
+++ b/pkg/ccl/oidcccl/BUILD.bazel
@@ -52,6 +52,7 @@ go_test(
         "//pkg/security/securitytest",
         "//pkg/security/username",
         "//pkg/server",
+        "//pkg/server/authserver",
         "//pkg/server/serverpb",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/sqlutils",
@@ -59,6 +60,8 @@ go_test(
         "//pkg/util/leaktest",
         "//pkg/util/log",
         "//pkg/util/randutil",
+        "@com_github_coreos_go_oidc//:go-oidc",
         "@com_github_stretchr_testify//require",
+        "@org_golang_x_oauth2//:oauth2",
     ],
 )

--- a/pkg/ccl/oidcccl/authentication_oidc.go
+++ b/pkg/ccl/oidcccl/authentication_oidc.go
@@ -123,10 +123,9 @@ var (
 //     manner, bypassing any password validation requirements, and redirect them to `/` so they can
 //     enjoy a logged-in experience in the Admin UI.
 type oidcAuthenticationServer struct {
-	mutex        syncutil.RWMutex
-	conf         oidcAuthenticationConf
-	oauth2Config oauth2.Config
-	verifier     *oidc.IDTokenVerifier
+	mutex   syncutil.RWMutex
+	conf    oidcAuthenticationConf
+	manager IOIDCManager
 	// enabled is used to store whether the user has flipped the enabled flag in the cluster settings
 	// if enabled is true and initialized is false, the code will continue to attempt to re-initialize
 	// the OIDC server every time a handler is invoked for the login or callback endpoints. This is
@@ -164,6 +163,93 @@ func (s *oidcAuthenticationServer) GetOIDCConf() ui.OIDCUIConf {
 
 		GenerateJWTAuthTokenEnabled: s.conf.generateJWTAuthTokenEnabled,
 	}
+}
+
+type oidcManager struct {
+	oauth2Config *oauth2.Config
+	verifier     *oidc.IDTokenVerifier
+}
+
+func (o *oidcManager) ExchangeVerifyGetClaims(
+	ctx context.Context, code string, idTokenKey string,
+) (map[string]json.RawMessage, error) {
+	credentials, err := o.Exchange(ctx, code)
+	if err != nil {
+		log.Errorf(ctx, "OIDC: failed to exchange code for token: %v", err)
+		return nil, err
+	}
+
+	rawIDToken, ok := credentials.Extra(idTokenKey).(string)
+	if !ok {
+		err := errors.New("OIDC: failed to extract ID token from the token credentials")
+		log.Error(ctx, "OIDC: failed to extract ID token from the token credentials")
+		return nil, err
+	}
+
+	idToken, err := o.Verify(ctx, rawIDToken)
+	if err != nil {
+		log.Errorf(ctx, "OIDC: unable to verify ID token: %v", err)
+		return nil, err
+	}
+
+	var claims map[string]json.RawMessage
+	if err := idToken.Claims(&claims); err != nil {
+		log.Errorf(ctx, "OIDC: unable to deserialize token claims: %v", err)
+		return nil, err
+	}
+
+	return claims, nil
+}
+
+func (o *oidcManager) Verify(ctx context.Context, s string) (*oidc.IDToken, error) {
+	return o.verifier.Verify(ctx, s)
+}
+
+func (o *oidcManager) Exchange(
+	ctx context.Context, s string, option ...oauth2.AuthCodeOption,
+) (*oauth2.Token, error) {
+	return o.oauth2Config.Exchange(ctx, s, option...)
+}
+
+func (o oidcManager) AuthCodeURL(s string, option ...oauth2.AuthCodeOption) string {
+	return o.oauth2Config.AuthCodeURL(s, option...)
+}
+
+type IOIDCManager interface {
+	Verify(context.Context, string) (*oidc.IDToken, error)
+	Exchange(context.Context, string, ...oauth2.AuthCodeOption) (*oauth2.Token, error)
+	AuthCodeURL(string, ...oauth2.AuthCodeOption) string
+	ExchangeVerifyGetClaims(context.Context, string, string) (map[string]json.RawMessage, error)
+}
+
+var _ IOIDCManager = &oidcManager{}
+
+var NewOIDCManager func(context.Context, oidcAuthenticationConf, string, []string) (IOIDCManager, error) = func(
+	ctx context.Context,
+	conf oidcAuthenticationConf,
+	redirectURL string,
+	scopes []string,
+) (IOIDCManager, error) {
+	provider, err := oidc.NewProvider(ctx, conf.providerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	oauth2Config := &oauth2.Config{
+		ClientID:     conf.clientID,
+		ClientSecret: conf.clientSecret,
+		RedirectURL:  redirectURL,
+
+		Endpoint: provider.Endpoint(),
+		Scopes:   scopes,
+	}
+
+	verifier := provider.Verifier(&oidc.Config{ClientID: conf.clientID})
+
+	return &oidcManager{
+		verifier:     verifier,
+		oauth2Config: oauth2Config,
+	}, nil
 }
 
 func reloadConfig(
@@ -221,15 +307,6 @@ func reloadConfigLocked(
 		return
 	}
 
-	provider, err := oidc.NewProvider(ctx, server.conf.providerURL)
-	if err != nil {
-		log.Warningf(ctx, "unable to initialize OIDC server, disabling OIDC: %v", err)
-		if log.V(1) {
-			log.Infof(ctx, "check provider URL OIDC cluster setting: "+OIDCProviderURLSettingName)
-		}
-		return
-	}
-
 	// Validation of the scope setting will require that we have the `openid` scope.
 	scopesForOauth := strings.Split(server.conf.scopes, " ")
 
@@ -242,16 +319,16 @@ func reloadConfigLocked(
 		return
 	}
 
-	server.oauth2Config = oauth2.Config{
-		ClientID:     server.conf.clientID,
-		ClientSecret: server.conf.clientSecret,
-		RedirectURL:  redirectURL,
-
-		Endpoint: provider.Endpoint(),
-		Scopes:   scopesForOauth,
+	manager, err := NewOIDCManager(ctx, server.conf, redirectURL, scopesForOauth)
+	if err != nil {
+		log.Warningf(ctx, "unable to initialize OIDC server, disabling OIDC: %v", err)
+		if log.V(1) {
+			log.Infof(ctx, "check provider URL OIDC cluster setting: "+OIDCProviderURLSettingName)
+		}
+		return
 	}
 
-	server.verifier = provider.Verifier(&oidc.Config{ClientID: server.conf.clientID})
+	server.manager = manager
 	server.initialized = true
 	log.Infof(ctx, "initialized OIDC server")
 }
@@ -364,30 +441,8 @@ var ConfigureOIDC = func(
 			return
 		}
 
-		credentials, err := oidcAuthentication.oauth2Config.Exchange(ctx, r.URL.Query().Get(codeKey))
+		claims, err := oidcAuthentication.manager.ExchangeVerifyGetClaims(ctx, r.URL.Query().Get(codeKey), idTokenKey)
 		if err != nil {
-			log.Errorf(ctx, "OIDC: failed to exchange code for token: %v", err)
-			http.Error(w, genericCallbackHTTPError, http.StatusInternalServerError)
-			return
-		}
-
-		rawIDToken, ok := credentials.Extra(idTokenKey).(string)
-		if !ok {
-			log.Error(ctx, "OIDC: failed to extract ID token from the token credentials")
-			http.Error(w, genericCallbackHTTPError, http.StatusInternalServerError)
-			return
-		}
-
-		idToken, err := oidcAuthentication.verifier.Verify(ctx, rawIDToken)
-		if err != nil {
-			log.Errorf(ctx, "OIDC: unable to verify ID token: %v", err)
-			http.Error(w, genericCallbackHTTPError, http.StatusInternalServerError)
-			return
-		}
-
-		var claims map[string]json.RawMessage
-		if err := idToken.Claims(&claims); err != nil {
-			log.Errorf(ctx, "OIDC: unable to deserialize token claims: %v", err)
 			http.Error(w, genericCallbackHTTPError, http.StatusInternalServerError)
 			return
 		}
@@ -479,7 +534,7 @@ var ConfigureOIDC = func(
 			return
 		}
 
-		credentials, err := oidcAuthentication.oauth2Config.Exchange(ctx, r.URL.Query().Get(codeKey))
+		credentials, err := oidcAuthentication.manager.Exchange(ctx, r.URL.Query().Get(codeKey))
 		if err != nil {
 			log.Errorf(ctx, "OIDC: failed to exchange code for token: %v", err)
 			log.Errorf(ctx, "%v", r.URL.Query().Get(codeKey))
@@ -498,7 +553,7 @@ var ConfigureOIDC = func(
 			rawToken = rawIDToken
 		}
 
-		token, err := oidcAuthentication.verifier.Verify(ctx, rawToken)
+		token, err := oidcAuthentication.manager.Verify(ctx, rawToken)
 		if err != nil {
 			log.Errorf(ctx, "OIDC: unable to verify ID token: %v", err)
 			http.Error(w, genericCallbackHTTPError, http.StatusInternalServerError)
@@ -667,7 +722,7 @@ var ConfigureOIDC = func(
 		}
 
 		http.SetCookie(w, kast.secretKeyCookie)
-		http.Redirect(w, r, oidcAuthentication.oauth2Config.AuthCodeURL(kast.signedTokenEncoded), http.StatusFound)
+		http.Redirect(w, r, oidcAuthentication.manager.AuthCodeURL(kast.signedTokenEncoded), http.StatusFound)
 	}))
 
 	reloadConfig(serverCtx, oidcAuthentication, locality, st)

--- a/pkg/ccl/oidcccl/authentication_oidc_test.go
+++ b/pkg/ccl/oidcccl/authentication_oidc_test.go
@@ -16,7 +16,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
-	"net/http/httptest"
 	"net/url"
 	"regexp"
 	"strings"
@@ -25,12 +24,15 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/server/authserver"
 	"github.com/cockroachdb/cockroach/pkg/server/serverpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/coreos/go-oidc"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/oauth2"
 )
 
 func TestOIDCBadRequestIfDisabled(t *testing.T) {
@@ -58,6 +60,36 @@ func TestOIDCBadRequestIfDisabled(t *testing.T) {
 	}
 }
 
+type mockOidcManager struct {
+	oauth2Config *oauth2.Config
+	claimEmail   string
+}
+
+func (m mockOidcManager) Verify(ctx context.Context, s string) (*oidc.IDToken, error) {
+	return nil, nil
+}
+
+func (m mockOidcManager) Exchange(
+	ctx context.Context, s string, option ...oauth2.AuthCodeOption,
+) (*oauth2.Token, error) {
+	return nil, nil
+}
+
+func (m mockOidcManager) AuthCodeURL(s string, option ...oauth2.AuthCodeOption) string {
+	return m.oauth2Config.AuthCodeURL(s, option...)
+}
+
+func (m mockOidcManager) ExchangeVerifyGetClaims(
+	ctx context.Context, s string, s2 string,
+) (map[string]json.RawMessage, error) {
+	x := map[string]json.RawMessage{}
+	// The email is surrounded by double quotes because it's a serialized JSON string.
+	x["email"] = json.RawMessage([]byte(fmt.Sprintf(`"%s"`, m.claimEmail)))
+	return x, nil
+}
+
+var _ IOIDCManager = &mockOidcManager{}
+
 func TestOIDCEnabled(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -67,89 +99,38 @@ func TestOIDCEnabled(t *testing.T) {
 	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 
-	// Set up a test OIDC server that serves the JSON discovery document
-	var issuer string
-	oidcHandler := func(w http.ResponseWriter, r *http.Request) {
-		if r.URL.Path == "/.well-known/openid-configuration" {
-			w.Header().Set("content-type", "application/json")
-			fakeDiscoveryDocument := `
-{
- "issuer": "` + issuer + `",
- "authorization_endpoint": "https://accounts.cockroachlabs.com/o/oauth2/v2/auth",
- "device_authorization_endpoint": "https://oauth2.cockroachlabsapis.com/device/code",
- "token_endpoint": "https://oauth2.cockroachlabsapis.com/token",
- "userinfo_endpoint": "https://openidconnect.cockroachlabsapis.com/v1/userinfo",
- "revocation_endpoint": "https://oauth2.cockroachlabsapis.com/revoke",
- "jwks_uri": "https://www.cockroachlabsapis.com/oauth2/v3/certs",
- "response_types_supported": [
-  "code",
-  "token",
-  "id_token",
-  "code token",
-  "code id_token",
-  "token id_token",
-  "code token id_token",
-  "none"
- ],
- "subject_types_supported": [
-  "public"
- ],
- "id_token_signing_alg_values_supported": [
-  "RS256"
- ],
- "scopes_supported": [
-  "openid",
-  "email",
-  "profile"
- ],
- "token_endpoint_auth_methods_supported": [
-  "client_secret_post",
-  "client_secret_basic"
- ],
- "claims_supported": [
-  "aud",
-  "email",
-  "email_verified",
-  "exp",
-  "family_name",
-  "given_name",
-  "iat",
-  "iss",
-  "locale",
-  "name",
-  "picture",
-  "sub"
- ],
- "code_challenge_methods_supported": [
-  "plain",
-  "S256"
- ],
- "grant_types_supported": [
-  "authorization_code",
-  "refresh_token",
-  "urn:ietf:params:oauth:grant-type:device_code",
-  "urn:ietf:params:oauth:grant-type:jwt-bearer"
- ]
-}`
-			_, _ = fmt.Fprint(w, fakeDiscoveryDocument)
-			return
+	usernameUnderTest := "test"
+
+	realNewManager := NewOIDCManager
+	NewOIDCManager = func(ctx context.Context, conf oidcAuthenticationConf, redirectURL string, scopes []string) (IOIDCManager, error) {
+		c := &oauth2.Config{
+			ClientID:     conf.clientID,
+			ClientSecret: conf.clientSecret,
+			RedirectURL:  redirectURL,
+
+			Endpoint: oauth2.Endpoint{
+				AuthURL: "https://provider.example.com/endpoint",
+			},
+			Scopes: scopes,
 		}
-		http.Error(w, "wrong path", http.StatusBadRequest)
+		return &mockOidcManager{oauth2Config: c, claimEmail: fmt.Sprintf("%s@example.com", usernameUnderTest)}, nil
 	}
-	testOIDCServer := httptest.NewServer(http.HandlerFunc(oidcHandler))
-	defer testOIDCServer.Close()
-	issuer = testOIDCServer.URL
+	defer func() {
+		NewOIDCManager = realNewManager
+	}()
 
 	// Set minimum settings to successfully enable the OIDC client
 	sqlDB := sqlutils.MakeSQLRunner(db)
-	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.provider_url = "`+testOIDCServer.URL+`"`)
+	sqlDB.Exec(t, fmt.Sprintf(`CREATE USER %s with password 'unused'`, usernameUnderTest))
+	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.provider_url = "providerURL"`)
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.client_id = "fake_client_id"`)
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.client_secret = "fake_client_secret"`)
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.redirect_url = "https://cockroachlabs.com/oidc/v1/callback"`)
+	sqlDB.Exec(t, `set cluster setting server.oidc_authentication.claim_json_key = "email"`)
+	sqlDB.Exec(t, `set cluster setting server.oidc_authentication.principal_regex = '^([^@]+)@[^@]+$'`)
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.oidc_authentication.enabled = "true"`)
 
 	testCertsContext := s.NewClientRPCContext(ctx, username.TestUserName())
-
 	client, err := testCertsContext.GetHTTPClient()
 	require.NoError(t, err)
 
@@ -158,44 +139,79 @@ func TestOIDCEnabled(t *testing.T) {
 		return http.ErrUseLastResponse
 	}
 
-	t.Run("login redirect", func(t *testing.T) {
-		resp, err := client.Get(s.AdminURL().WithPath("/oidc/v1/login").String())
-		if err != nil {
-			t.Fatalf("could not issue GET request to admin server: %s", err)
-		}
-		defer resp.Body.Close()
+	resp, err := client.Get(s.AdminURL().WithPath("/oidc/v1/login").String())
+	if err != nil {
+		t.Fatalf("could not issue GET request to admin server: %s", err)
+	}
+	defer resp.Body.Close()
 
-		if resp.StatusCode != 302 {
-			t.Fatalf("expected 302 status code but got: %d", resp.StatusCode)
-		}
-		if resp.Cookies()[0].Name != secretCookieName {
-			t.Fatal("Missing cookie")
-		}
-		authURL, err := url.Parse(resp.Header.Get("Location"))
-		require.NoError(t, err)
-		if authURL.Query().Get("client_id") != "fake_client_id" {
-			t.Fatal("expected fake client_id", authURL)
-		}
-		const expectedRedirectURL = "https://cockroachlabs.com/oidc/v1/callback"
-		if authURL.Query().Get("redirect_uri") != expectedRedirectURL {
-			t.Fatal("expected fake redirect_url", authURL)
-		}
+	if resp.StatusCode != 302 {
+		t.Fatalf("expected 302 status code but got: %d", resp.StatusCode)
+	}
+	if resp.Cookies()[0].Name != secretCookieName {
+		t.Fatal("Missing cookie")
+	}
+	cookie := resp.Cookies()[0]
 
-		state, err := decodeOIDCState(authURL.Query().Get("state"))
-		require.NoError(t, err)
-		// If we use hmac.Sum with the Message, it gets prepended to the hash.
-		if strings.Contains(string(state.TokenMAC), string(state.Token)) {
-			t.Fatal("HMAC generated incorrectly.")
-		}
+	authURL, err := url.Parse(resp.Header.Get("Location"))
+	require.NoError(t, err)
+	if authURL.Query().Get("client_id") != "fake_client_id" {
+		t.Fatal("expected fake client_id", authURL)
+	}
+	const expectedRedirectURL = "https://cockroachlabs.com/oidc/v1/callback"
+	if authURL.Query().Get("redirect_uri") != expectedRedirectURL {
+		t.Fatal("expected fake redirect_url", authURL)
+	}
 
-		key, err := base64.URLEncoding.DecodeString(resp.Cookies()[0].Value)
-		require.NoError(t, err)
-		mac := hmac.New(sha256.New, key)
-		mac.Write(state.Token)
-		if !hmac.Equal(mac.Sum(nil), state.TokenMAC) {
-			t.Fatal("HMAC hash doesn't match TokenMAC")
+	stateParam := authURL.Query().Get("state")
+	state, err := decodeOIDCState(stateParam)
+	require.NoError(t, err)
+	// If we use hmac.Sum with the Message, it gets prepended to the hash.
+	if strings.Contains(string(state.TokenMAC), string(state.Token)) {
+		t.Fatal("HMAC generated incorrectly.")
+	}
+
+	key, err := base64.URLEncoding.DecodeString(resp.Cookies()[0].Value)
+	require.NoError(t, err)
+	mac := hmac.New(sha256.New, key)
+	mac.Write(state.Token)
+	if !hmac.Equal(mac.Sum(nil), state.TokenMAC) {
+		t.Fatal("HMAC hash doesn't match TokenMAC")
+	}
+
+	// Simulate the OIDC provider invoking our callback.
+	req, err := http.NewRequest("GET", s.AdminURL().WithPath("/oidc/v1/callback").String(), nil)
+	require.NoError(t, err)
+
+	// The cookie set on the login request must be retained by the
+	// browser when the callback is invoked in order to verify that the
+	// `state` param matches the one that was sent during login.
+	req.AddCookie(cookie)
+	q := req.URL.Query()
+	q.Add("state", stateParam)
+	req.URL.RawQuery = q.Encode()
+
+	resp, err = client.Do(req)
+	if err != nil {
+		t.Fatalf("could not issue GET request to callback endpoint: %s", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 307 {
+		t.Fatalf("expected 307 status code but got: %d", resp.StatusCode)
+	}
+	if resp.Header.Get("Location") != "/" {
+		t.Fatalf("expected to be redirected to root")
+	}
+	foundCookie := false
+	for _, c := range resp.Cookies() {
+		if c.Name == authserver.SessionCookieName {
+			foundCookie = true
 		}
-	})
+	}
+	if !foundCookie {
+		t.Fatalf("no session cookie found in callback response")
+	}
 }
 
 func TestKeyAndSignedTokenIsValid(t *testing.T) {

